docs/designs/concepts.md

```diff
@@ -0,0 +1,165 @@
+# Concepts Design Doc
+
+This document describes core concepts used in ElasticDL, and discusses their representations.
+
+## Core Concepts
+
+Fundamentally, machine learning is using algorithms to extract information from raw data and represent it in some type of model.
+
+There two kinds of parameters in a model, one is dense parameter, the other is sparse parameter. The sparse parameter is used by the embedding layer, so it is also called embedding table.
+
+Both dense parameter and embedding table includes a tensor data field, and some other auxiliary fields. The auxiliary fields indluce a name. Embedding table has a extra auxiliary field, indices field.
+
+Tensor is used to represented a n-dim data. There are many data types, such as int8/float32.
+
+Each kinds of parameter has its own kind of gradient. Dense parameter has dense parameter gradient, embedding table has embedding table gradient.
+
+Let's make a short summary, following is all the core concepts used in ElasticDL:
+
+- Model
+- DenseParam
+- EmbeddingTable
+- Tensor
+- DenseParamGradient
+- EmbeddingTableGradient
+
+## In-Memory Representation
+
+```go
+type TensorDtype = int64
+
+const (
+    Invalid TensorDtype = iota
+    Int8 
+    Int16
+    Int32
+    Int64
+    Float16
+    Float32
+    Float64
+    Bool
+)
+
+type Buffer struct {
+    Data   []byte
+    Length int64
+    Dtype  TensorDtype
+}
+
+type Tensor struct {
+    Content *Buffer
+    Dims    []int64
+    Indices []int64
+}
+
+type DenseParam struct {
+    Name   string
+    Tensor Tensor
+}
+
+type Emdedding = Buffer // alias for better understanding
+
+type EmbeddingTable struct {
+    Name       string
+    Embeddings map[string]*Emdedding
```

Comment:
The key of Embeddings is int64. The embeddings maps from an item id to an embedding vector.